Improve verification results

jku opened the pull request:
This is an attempt at fixing the verification result handling that has been found to not work well. This API is useful mostly to repository implementations: clients should likely keep using `verify_delegate() `

**Modify VerificationResult**
* Remove `union()`: this is now handled by the new RootVerificationResult
* include threshold: This is useful and can be added now that union() does not exist
* Use `dict[str, Key]` instead of `set[str]` for signed and unsigned
* Note that the (broken metadata) case of role keyid that does not have a matching key, is now not counted as "unsigned" as it was before: it's just not counted at all 

**Add RootVerificationResult**
* Externally looks like VerificationResult (without threshold)
* It is actually two VerificationResults in a trenchcoat: `Root.get_root_verification_result()` is provided as a straightforward way to get one

Note that `RootVerificationResult.signed` and `RootVerificationResult.unsigned` assume that there are no keyid collisions between the two verifying roles (same keyid is not used for two different keys): This does **not** affect `verified` status in any way (only the `signed` and `unsigned` lists) so I think this is totally acceptable.

So far the usage looks quite nice but I think we want to try to use this in an app or two to avoid redoing the mistake of the first VerificationResult... 

Fixes #2544 

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

jku committed:
Metadata API: Refactor VerificationResult

This is an API break as VerificationResult changes:
 * Now contains threshold
 * Now contains Keys and not just keyids

Note that there is a small edge case functionality change:
 * if the role does not have a key for the keyid, then we no longer
   include that key in "unsigned"

I think that is an acceptable change.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku committed:
Metadata API: Implement RootVerificationResult

This is a thin wrapper over two VerificationResults:
useful when verifying root signatures.

Now the API for getting verification results for root and
the API for getting the results for other metadata is different.

Client use cases can continue using verify_delegate() so should not
be affected.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku committed:
tests: Update to new VerificationResult

Changes are
* expected result changes (like the handling of keyids without keys)
* test refactoring to have access to the Key
* Removal of union test
* use the fact that VerificationResult is Truthy in asserts
  (to get 1 more line of coverage)

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku committed:
tests: Add tests for root verification

This does much the same tests as test_signed_get_verification_result()
above it does, just using two root roles.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku commented:
Still wondering if it would  be useful for the two result types to have a common base type, instead of being just duck typed to feel similar. 

The way I see a signing tool or repository UI using these is this:
* getting verification results happens separately for root and other roles (because it has to)
* but printing the results should be possible with the same code (with maybe a small tweak for root case)

So maybe it is fine for the app to have something like
```
def describe_result(result: VerificationResult | RootVerificationResult) -> str:
    ...
```

I would maybe rather not create some complicated inheritance (and inevitably large classes) when the dataclasses look like they might just work like that...

jku commented on tuf/api/metadata.py:
```diff
@@ -653,30 +652,61 @@ class VerificationResult:
+    @property
+    def signed(self) -> Dict[str, Key]:
+        """Dictionary of all signing keys that have signed, from both VerificationResults"""
+        return self.first.signed | self.second.signed
```
this unfortunately requires python 3.9

Following likely does the same with dict unpacking but is even less readable
`return {**self.first.signed, **self.second.signed}`

jku committed:
Metadata API: Workaround for Python <3.9

dict unions are only supported in 3.9.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku committed:
Metadata API: Simplify verify_delegate()

Now that VerificationResult has threshold, this can be simpler.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -935,6 +963,34 @@ def get_key(self, keyid: str) -> Key:  # noqa: D102
+
+        Verify root metadata with two roles (the root role from `self` and
+        `other`). If you have only one role (in the case of root v1) you can
+        provide the same Root as both `self` and `other`.
```
I like this. It treats v1 as the edge case which it is. So far in my code it always looked like the non-v1 case was the special case.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -935,6 +963,34 @@ def get_key(self, keyid: str) -> Key:  # noqa: D102
+        """
+        return RootVerificationResult(
+            self.get_verification_result(Root.type, payload, signatures),
+            other.get_verification_result(Root.type, payload, signatures),
```
Is it premature optimisation to call get_verification_result only once, if self==other?

jku replied:
it probably is

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -653,30 +652,63 @@ class VerificationResult:
+    """Signature verification result for root metadata.
+
+    Root must be verified by itself and the previous root version. This
+    dataclass represents that combination. For the edge case of first version
+    of root, the same VerificationResult can be used twice.
```
I had to read `get_root_verification_result` to understand what this means. Maybe rephrase to something like:


```suggestion
    dataclass represents both results. For the edge case of first version
    of root, the results are identical.
```

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -653,30 +652,63 @@ class VerificationResult:
-        signed: Set of delegated keyids, which validly signed.
-        unsigned: Set of delegated keyids, which did not validly sign.
+        threshold: Number of required signatures.
+        signed: dict of keyid:Key containing the keys that have signed.
+        unsigned: dict of keyid:Key containing the keys that have not signed.
```
nit: Would you mind to change this?

```suggestion
        signed: dict of keyids to Keys that have signed.
        unsigned: dict of keyids to Keys that have not signed.
```

The second colon throws off the parser in my head. 😆

lukpueh reviewed (commented):
This looks very helpful. Let me try it in RSTUF and see if it really is.

jku committed:
examples: Use verification results in repo example

This is an example of using the verification resutls in a repository.

The only remaining tricky part is in _get_verification_result():
* has to figure out the delegating metadata (something we currently
  cannot provide in repository.Repository for the general case)
* Needs a special case for first root

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -935,6 +963,34 @@ def get_key(self, keyid: str) -> Key:  # noqa: D102
 
+    def get_root_verification_result(
+        self,
+        other: "Root",
```
It would be convenient to allow passing `None` for `other` and default to `self`. 

Plus we could check if the versions are correct:
```python
if self.version == 1 and other.version != 1 or self.version > 1 and other.version != self.version - 1:
   raise ValueError 
```

lukpueh replied:
Similarly, if we decide to make the interface version-aware, I suggest to rename `other` to `previous`.

jku replied:
that sounds reasonable

jku replied:
oops, I didn't do any version checking yet. I will add some of it...

jku replied:
Right, I added the check for sequential  versions. This seems uncontroversial.

I didn't check that "initial version" is v1. I don't think that's a spec requirement and I could even  imagine some semi-practical cases where the opposite could happen

lukpueh commented:
I tried this in RSTUF and found it very helpful.

I think adding a common base class would be okay. The inheritance model could look something like this.
```python
class VerificationResult:
  verified: bool
  signed: dict[str, Key]
  unsigned: dict[str, Key]

class Single(VerificationResult):
  threshold: int

class Combined(VerificationResult):
  first: Single
  second: Single
```

But this does not add much value. The more interesting question is, if we want to provide other common behavior. Two things that I would have found useful:

- a list of missing signature infos:
   unique tuples of `threshold -  len(signed)` and `unsigned` if `verified` is `False` per Single result

- keys that effectively decrease the missing signature count
  union of `unsigned` if `verified` is `False` per Single result

jku committed:
Metadata API: Tweak get_root_verification_result args

Change the "other" argument to optional "previous" and
handle the None case in code.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku committed:
Metadata API: Improve docs for RootVerificationResult

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku commented:
The common base class probably looks more complicated: because Combined wants to override `verified`, `signed` and `unsigned` they probably would have to be properties in the parent class already.

jku commented:
> * a list of missing signature infos:
> unique tuples of threshold -  len(signed) and unsigned if verified is False per Single result
> * keys that effectively decrease the missing signature count
> union of unsigned if verified is False per Single result

I'm not completely sold on these because :
* this is only relevant for root and only when root signers change
* I think most repositories in practice would want to encourage everyone to sign (to make the signing events Proof-of-Ownership events at the same time) 
* Handling multliple tuples does not seem any easier than handling `result.first` and `result.second` (or if you think it is we can just make those a tuple instead of two fields) so you could just handle the underlying VerificationResults yourself

We could add this to `VerificationResult` if it makes things easier:
```
@property
def missing(self) -> int:
    return max(0, self.threshold - len(self.signed))
```

jku committed:
Metadata API: Make sanity checks in root verification

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku committed:
Metadata API: Add VerificationResult.missing

This is helper to tell how many signatures are still required.
Also change the order of Roots given to RootVerificationResult
(this way first is version N, second is version N+1).

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku committed:
tests: Update the root verification tests

Change tests so the previous root version is what the code expects.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku committed:
tests: Add VerificationResult tests

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -719,33 +756,30 @@ def get_verification_result(
             try:
                 key = self.get_key(keyid)
             except ValueError:
-                unsigned.add(keyid)
```
Interesting. This means that in the new VerificationResult there might be role keyids that are neither in signed nor unsigned. That's probably okay. But should we mention it somewhere. Maybe in the PR description?

jku replied:
Yeah it is a change -- I think I mentioned it in the first commit message , can add to PR decription.

 I was also considering just raising at this point (because clearly the role is in a bad state)... but then that sounds like punishing the delegated role signing  for delegator metadata error -- an error that can still leave the delegated role capable of signing.
